File: reversi/reversi.py
FIELD_WIDTH = 8


def is_valid_index(cell_idx):
    if 0 < cell_idx[0] < FIELD_WIDTH and 0 < cell_idx[1] < FIELD_WIDTH:
        return True

    return False


class InvalidCell(Exception):
    def __init__(self, *args):
        if args:
            self.message = args[0]
        else:
            self.message = None

    def __str__(self):
        if self.message:
            return f"InvalidCell: {self.message}"
        else:
            return "InvalidCell"


class Reversi():
    def __init__(self):
        self.is_finished = False
        self.field = [[' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', 'w', 'b', ' ', ' ', ' '],
                      [' ', ' ', ' ', 'b', 'w', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ']]

    def inverseof(self, color):
        if color == 'b':
            return 'w'
        else:
            return 'b'

    def get_valid_neighbours(self, cell_idx):
        result = []
        neighbours = [(cell_idx[0]-1, cell_idx[1]),
                      (cell_idx[0]-1, cell_idx[1]+1),
                      (cell_idx[0], cell_idx[1]+1),
                      (cell_idx[0]+1, cell_idx[1]+1),
                      (cell_idx[0]+1, cell_idx[1]),
                      (cell_idx[0]+1, cell_idx[1]-1),
                      (cell_idx[0], cell_idx[1]-1),
                      (cell_idx[0]-1, cell_idx[1]-1)]
        for n in neighbours:
            if is_valid_index(n):
                result.append(n)

        return result

    def get_inverse_neighbours(self, cell_idx):
        result = []
        inverse = self.inverseof(self.field[cell_idx[0]][cell_idx[1]])
        neighbours = self.get_valid_neighbours(cell_idx)
        for n in neighbours:
            neighbour_cell = self.field[n[0]][n[1]]
            if neighbour_cell == inverse:
                result.append(n)

        return result

    def get_vectors(self, cell_idx, points):
        result = []
        for p in points:
            result.append((p[0] - cell_idx[0], p[1] - cell_idx[1]))

        return result

    def find_empty_cell(self, cell_idx, direction):
        start = self.field[cell_idx[0]][cell_idx[1]]
        inverse = self.inverseof(start)
        y = cell_idx[0] + direction[0]
        x = cell_idx[1] + direction[1]
        while True:
            if not is_valid_index((y, x)):
                return None
            current = self.field[y][x]
            if current == inverse:
                return None
            if current == ' ':
                return (y, x)
            if current == start:
                y += direction[0]
                x += direction[1]
                continue

    def get_available_moves(self, cell_idx):
        available_moves = []
        n = self.get_inverse_neighbours(cell_idx)
        d = self.get_vectors(cell_idx, n)
        for i in range(len(n)):
            move = self.find_empty_cell(n[i], d[i])
            if move is not None:
                available_moves.append(move)

        return available_moves

    def get_pieces(self, color):
        pieces = []
        for i in range(FIELD_WIDTH):
            for j in range(FIELD_WIDTH):
                current = self.field[i][j]
                if current == color:
                    pieces.append((i, j))

        return pieces

    def get_coverage(self, color):
        pieces = self.get_pieces(color)
        coverage = {}

        for p in pieces:
            moves = self.get_available_moves(p)
            for m in moves:
                if m not in coverage:
                    coverage[m] = [p]
                else:
                    coverage[m].append(p)

        return coverage

    def make_turn(self, cell_idx, color):
        coverage = self.get_coverage(color)
        self.flip_pieces(cell_idx, coverage[cell_idx])

File: src/reversi.nim
import tables, sequtils

const FieldSize* = 8
const Empty* = ' '
const Black* = 'b'
const White* = 'w'


type CellIndex* = tuple
    y: int
    x: int

const InvalidCell = (-1, -1)


type Direction = tuple
    y: int
    x: int


type Coverage* = Table[CellIndex, seq[CellIndex]]


type Reversi* = ref object
    field: array[FieldSize, array[FieldSize, char]]
    blackHole: CellIndex

proc newReversi*(blackHole: CellIndex): Reversi =
    let field = [[Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty],
                 [Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty],
                 [Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty],
                 [Empty, Empty, Empty, White, Black, Empty, Empty, Empty],
                 [Empty, Empty, Empty, Black, White, Empty, Empty, Empty],
                 [Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty],
                 [Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty],
                 [Empty, Empty, Empty, Empty, Empty, Empty, Empty, Empty]]
    return Reversi(field: field, blackHole: blackHole)


proc inverseof(color: char): char =
    if color == Black:
        return White
    return Black


proc isValidIndex(cellIndex: CellIndex): bool =
    if (0 <= cellIndex.y and 
        cellIndex.y < FieldSize and 
        0 <= cellIndex.x and 
        cellIndex.x < FieldSize):
        return true

    return false


proc getValidNeighbours(cellIndex: CellIndex): seq[CellIndex] =
    let neighbours = [(cellIndex.y - 1, cellIndex.x),
                      (cellIndex.y - 1, cellIndex.x + 1),
                      (cellIndex.y, cellIndex.x + 1),
                      (cellIndex.y + 1, cellIndex.x + 1),
                      (cellIndex.y + 1, cellIndex.x),
                      (cellIndex.y + 1, cellIndex.x - 1),
                      (cellIndex.y, cellIndex.x - 1),
                      (cellIndex.y - 1, cellIndex.x - 1)]
    for n in neighbours:
        if isValidIndex(n):
            result.add(n)


proc getInverseNeighbours(self: Reversi, cellIndex: CellIndex): seq[CellIndex] =
    let inverse = inverseof(self.field[cellIndex.y][cellIndex.x])
    let neighbours = getValidNeighbours(cellIndex)
    for n in neighbours:
        let neighbour_cell = self.field[n[0]][n[1]]
        if neighbour_cell == inverse:
            result.add(n)


proc findEmptyCell(self: Reversi, cellIndex: CellIndex, direction: Direction): CellIndex =
    let color = self.field[cellIndex.y][cellIndex.x]
    let inverse = inverseof(color)
    var y = cellIndex.y + direction.y
    var x = cellIndex.x + direction.x
    while true:
        if not isValidIndex((y, x)):
            return InvalidCell
        let current = self.field[y][x]
        if current == inverse:
            return InvalidCell
        if current == Empty:
            return (y, x)
        if current == color:
            y += direction.y
            x += direction.x
            continue


proc getDirections(cellIndex: CellIndex, points: seq[CellIndex]): seq[Direction] =
    for p in points:
        result.add((p[0] - cellIndex.y, p[1] - cellIndex.x))


proc getAvailableMoves(self: Reversi, cellIndex: CellIndex): seq[CellIndex] =
    let n = self.getInverseNeighbours(cellIndex)
    let d = getDirections(cellIndex, n)
    for i in 0..<len(n):
        let move = self.findEmptyCell(n[i], d[i])
        if move != InvalidCell:
            result.add(move)

proc getPieces(self: Reversi, color: char): seq[CellIndex] =
    for i in 0..<FieldSize:
        for j in 0..<FieldSize:
            let current = self.field[i][j]
            if current == color:
                result.add((i, j))


proc getCoverageWithBlackHole(self: Reversi, color: char): Coverage =
    let pieces = self.getPieces(color)

    for p in pieces:
        let moves = self.getAvailableMoves(p)
        for m in moves:
            if not result.hasKey(m):
                result[m] = @[p]
            else:
                result[m].add(p)


proc getCoverage*(self: Reversi, color: char): Coverage =
    result = self.getCoverageWithBlackHole(color)
    if result.hasKey(self.black_hole):
        result.del(self.black_hole)


proc flipRow(self: Reversi, cellIndex: CellIndex, direction: Direction) =
    let color = self.field[cellIndex.y][cellIndex.x]
    let inverse = inverseof(color)
    var y = cellIndex.y + direction.y
    var x = cellIndex.x + direction.x
    while true:
        let current = self.field[y][x]
        if current == inverse:
            self.field[y][x] = color
            y += direction.y
            x += direction.x
            continue
        break


proc normalizeDirection(direction: Direction): Direction = 
    let y = direction.y 
    let x = direction.x
    let ly = abs(y)
    let lx = abs(x)
    if ly > lx:
        return (int(y / ly), int(x / ly))
    if x == 0:
        return (y, x)

    return (int(y / lx), int(x / lx))


proc flipPieces(self: Reversi, attacked: CellIndex, attackers: seq[CellIndex]) =
    let color = self.field[attackers[0].y][attackers[0].x]
    self.field[attacked.y][attacked.x] = color
    let directions = map(getDirections(attacked, attackers), normalizeDirection)
    for d in directions:
        self.flipRow(attacked, d)


proc isFinished*(self: Reversi): bool =
    let blackCoverage = self.getCoverage(Black)
    let whiteCoverage = self.getCoverage(White)
    if len(blackCoverage) == 0 and len(whiteCoverage) == 0:
        return true
    return false


proc makeTurn*(self: Reversi, cellIndex: CellIndex, color: char) =
    let coverage = self.getCoverage(color)
    self.flipPieces(cellIndex, coverage[cellIndex])


proc calculateScore*(self: Reversi): (int, int) =
    var black = len(self.getPieces(Black))
    var white = len(self.getPieces(White))

    return (black, white)
